File: app/src/main/java/com/aura/features/home/data/ConversationRepository.kt
package com.aura.features.home.data

import android.content.Context
import android.util.Log
import com.aura.features.home.domain.ConversationMode
import com.aura.features.home.domain.ConversationStatus
import com.aura.features.home.domain.TranscriptMessage
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.withContext
import javax.inject.Inject
import javax.inject.Singleton

private const val TAG = "ConversationRepository"

@Singleton
class ConversationRepository @Inject constructor(
    private val tokenDataSource: ConversationTokenRemoteDataSource,
    private val sessionDataSource: ElevenLabsSessionDataSource,
) {
    val status: StateFlow<ConversationStatus> = sessionDataSource.status
    val mode: StateFlow<ConversationMode> = sessionDataSource.mode
    val isMuted: StateFlow<Boolean> = sessionDataSource.isMuted
    val transcript: StateFlow<List<TranscriptMessage>> = sessionDataSource.transcript

    suspend fun fetchToken(): String = tokenDataSource.fetchToken()

    suspend fun startConversation(context: Context, token: String): Result<Unit> = withContext(
        Dispatchers.IO
    ) {
        runCatching {
            sessionDataSource.start(token, context)
        }.onFailure { Log.e(TAG, "startConversation failed", it) }
    }

    fun endConversation() = sessionDataSource.stop()

    fun setMicMuted(muted: Boolean) = sessionDataSource.setMicMuted(muted)
}

File: mobile/lib/features/home/domain/repositories/conversation_repository.dart
import 'package:aura/core/errors/result.dart';
import '../entities/conversation_entity.dart';
import '../entities/transcript_message_entity.dart';

abstract class ConversationRepository {
  Stream<ConversationStatus> get statusStream;
  Stream<ConversationMode> get modeStream;
  Stream<bool> get isMutedStream;
  Stream<List<TranscriptMessageEntity>> get transcriptStream;

  Future<Result<String>> fetchToken();
  Future<Result<void>> startConversation(String token);
  Future<Result<void>> stopConversation();
  Future<Result<void>> setMicMuted(bool muted);
}
